File: python/src/smooai_audit/schema.py
"""Canonical audit event schema (Python port of @smooai/audit).

Mirrors the TypeScript ``auditEventSchema`` field-for-field. Verified
byte-for-byte against the shared parity corpus.

TODO(audit-impl): finalize the field set against the parity corpus.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class AuditEvent(BaseModel):
    """A single audit event — the shared shape every language SDK emits."""

    id: str
    org_id: str
    timestamp: str
    actor: str
    action: str
    resource: str | None = None
    metadata: dict[str, Any] | None = None
    previous_hash: str | None = None

File: rust/audit/src/schema.rs
//! Canonical audit event schema (Rust port of `@smooai/audit`).
//!
//! Mirrors the TypeScript `auditEventSchema` field-for-field and is verified
//! byte-for-byte against the shared parity corpus.
//!
//! TODO(audit-impl): finalize the field set against the parity corpus.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

/// A single audit event — the shared shape every language SDK emits.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AuditEvent {
    /// Stable unique id for this event (UUID).
    pub id: String,
    /// Organization the event belongs to — hash chains are per-org-per-day.
    pub org_id: String,
    /// RFC 3339 / ISO-8601 UTC timestamp.
    pub timestamp: String,
    /// Who performed the action.
    pub actor: String,
    /// What happened, e.g. `"record.delete"`.
    pub action: String,
    /// The thing acted upon.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resource: Option<String>,
    /// Arbitrary structured context. Must serialize canonically.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Map<String, Value>>,
    /// Hex SHA-256 of the previous event in this org/day chain (`""` for genesis).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub previous_hash: Option<String>,
}
